# Lab 7 (Week 2 — Lab 2): Regression Models

**Name:** Numair Fahad

**Role:** AI/ML Intern

**Company:** Zynvex Solutions

**GitHub:** https://github.com/numair-2003/AIML-Internship-NumairFahad

Welcome to Lab 7! In this lab, you will build your first predictive models: **Regression Models**, which predict continuous numeric values (such as prices, scores, or measurements) rather than categories.

**For detailed theory on Linear Regression, Cost Functions (MSE), Multiple Regression, Polynomial Regression, and Evaluation Metrics (MSE, RMSE, R²), see the Lab 7 PDF and Learning Resources.**

This notebook focuses on the practical, hands-on workflow using **scikit-learn** and **Pandas**.

---
## Dataset Overview: California Housing Dataset

In this lab, we will use the **California Housing Dataset** (available directly in `scikit-learn`).

- **Target variable (`y`)**: `MedHouseVal` — Median house value for California districts (expressed in hundreds of thousands of dollars, e.g., 2.0 = $200,000).
- **Features (`X`)**:
  1. `MedInc`: Median income in block group
  2. `HouseAge`: Median house age in block group
  3. `AveRooms`: Average number of rooms per household
  4. `AveBedrms`: Average number of bedrooms per household
  5. `Population`: Block group population
  6. `AveOccup`: Average number of household members
  7. `Latitude`: Block group latitude
  8. `Longitude`: Block group longitude

## Example 1: Loading & Exploring the Dataset

First, let's load the dataset, create a Pandas DataFrame, and inspect the target variable and features.

In [5]:
from sklearn.datasets import fetch_california_housing
import pandas as pd
import numpy as np

# Load California Housing dataset
housing = fetch_california_housing(as_frame=True)
df = housing.frame

print("Dataset shape:", df.shape)
print("\nFirst 5 rows:")
print(df.head())

Dataset shape: (20640, 9)

First 5 rows:
   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedHouseVal  
0    -122.23        4.526  
1    -122.22        3.585  
2    -122.24        3.521  
3    -122.25        3.413  
4    -122.25        3.422  


## Example 2: Simple Linear Regression & Model Evaluation

Let's predict house prices (`MedHouseVal`) using just one feature: **Median Income (`MedInc`)**.

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Select single feature and target
X_simple = df[['MedInc']]
y = df['MedHouseVal']

# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_simple, y, test_size=0.2, random_state=42)

# Train model
model_simple = LinearRegression()
model_simple.fit(X_train, y_train)

# Predictions
y_pred_simple = model_simple.predict(X_test)

# Evaluation
mse = mean_squared_error(y_test, y_pred_simple)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred_simple)

print(f"Slope (Coefficient): {model_simple.coef_[0]:.4f}")
print(f"Intercept: {model_simple.intercept_:.4f}")
print(f"MSE: {mse:.4f}")
print(f"RMSE: {rmse:.4f}")
print(f"R2 Score: {r2:.4f}")

Slope (Coefficient): 0.4193
Intercept: 0.4446
MSE: 0.7091
RMSE: 0.8421
R2 Score: 0.4589


## Example 3: Multiple Linear Regression

Now, let's use multiple features (`MedInc`, `HouseAge`, `AveRooms`) to improve our prediction model.

In [7]:
# Select multiple features
features = ['MedInc', 'HouseAge', 'AveRooms']
X_multi = df[features]

X_train_m, X_test_m, y_train_m, y_test_m = train_test_split(X_multi, y, test_size=0.2, random_state=42)

# Train multiple linear regression
model_multi = LinearRegression()
model_multi.fit(X_train_m, y_train_m)

y_pred_multi = model_multi.predict(X_test_m)

# Evaluate
rmse_multi = np.sqrt(mean_squared_error(y_test_m, y_pred_multi))
r2_multi = r2_score(y_test_m, y_pred_multi)

print(f"Multiple Regression RMSE: {rmse_multi:.4f}")
print(f"Multiple Regression R2 Score: {r2_multi:.4f}")
print("\nFeature Coefficients:")
for feature, coef in zip(features, model_multi.coef_):
    print(f"{feature}: {coef:.4f}")

Multiple Regression RMSE: 0.8117
Multiple Regression R2 Score: 0.4972

Feature Coefficients:
MedInc: 0.4448
HouseAge: 0.0168
AveRooms: -0.0281


## Practice Exercise 1: Predicting with a Custom Inputs

**Exercise:** Use the trained `model_simple` from Example 2 to predict the house value (`MedHouseVal`) for a neighborhood with a median income (`MedInc`) of **5.0** (which represents $50,000).

In [12]:
### YOUR CODE HERE ###
prediction = model_simple.predict(pd.DataFrame([[5.0]], columns=['MedInc']))
print(f"Predicted median house value for a median income of 5.0: {prediction[0]:.4f}")
### END ###

Predicted median house value for a median income of 5.0: 2.5413


## Practice Exercise 2: Polynomial Regression

**Exercise:** Perform Polynomial Regression of **degree=2** on the single feature `MedInc`. Train the model and calculate the R² Score on the test set.

In [14]:
from sklearn.preprocessing import PolynomialFeatures

### YOUR CODE HERE ###
# 1. Create PolynomialFeatures with degree 2
poly = PolynomialFeatures(degree=2)

# 2. Transform X_train and X_test
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# 3. Fit LinearRegression on transformed data
model = LinearRegression()
model.fit(X_train_poly, y_train)

# 4. Calculate and print R2 Score
y_pred_poly = model.predict(X_test_poly)
r2_poly = r2_score(y_test, y_pred_poly)

print(f"Polynomial Regression R² Score: {r2_poly:.4f}")
### END ###

Polynomial Regression R² Score: 0.4633


## Lab Tasks

Complete the following tasks in this notebook below this cell.

1. **Full Feature Regression**: Train a Linear Regression model using ALL 8 features from the California Housing dataset. Perform an 80-20 train-test split. Print the RMSE and R² score.
2. **Coefficient Analysis**: Display all feature names alongside their learned coefficients for the model built in Task 1. Identify which feature has the strongest positive impact on house prices.
3. **Comparison**: Compare the R² score of the simple regression model (Example 2) with the full feature model (Task 1). In a Markdown cell, write 2-3 sentences explaining why including more features improved or changed the performance.
4. **Reflection**: Write 2-3 sentences on what you learned in this lab and how evaluation metrics (RMSE vs R²) help us understand model performance.

In [16]:
# Task 1
X_full = df.drop('MedHouseVal', axis=1)
y = df['MedHouseVal']

X_train_full, X_test_full, y_train_full, y_test_full = train_test_split(X_full, y, test_size=0.2, random_state=42)

model_full = LinearRegression()
model_full.fit(X_train_full, y_train_full)

y_pred_full = model_full.predict(X_test_full)

rmse_full = np.sqrt(mean_squared_error(y_test_full, y_pred_full))
r2_full = r2_score(y_test_full, y_pred_full)

print(f"Full Feature Regression RMSE: {rmse_full:.4f}")
print(f"Full Feature Regression R² Score: {r2_full:.4f}")

Full Feature Regression RMSE: 0.7456
Full Feature Regression R² Score: 0.5758


In [18]:
# Task 2
print("Feature Learned Coefficients:")
for feature_full, coefficient_full in zip(X_full.columns, model_full.coef_):
  print(f"{feature_full}: {coefficient_full:.4f}")

max_coef_index = np.argmax(model_full.coef_)
print(f"\nFeature having the strongest positive impact on house prices: {X_full.columns[max_coef_index]}")

Feature Learned Coefficients:
MedInc: 0.4487
HouseAge: 0.0097
AveRooms: -0.1233
AveBedrms: 0.7831
Population: -0.0000
AveOccup: -0.0035
Latitude: -0.4198
Longitude: -0.4337

Feature having the strongest positive impact on house prices: AveBedrms


# Task 3

The **simple linear regression model**, where **only MedInc** feature was used, achieved an **R² score** of **0.4589**, whereas the **full feature model** achieved a relatively high **R² score** of **0.5758**. This infers that including more features improved the model’s ability to explain the **distribution in house prices**. However, the improvement is not extremely large, which suggests that **Median Income** is already a very strong predictor on its own.

# Task 4

In this lab, I learned how to build and evaluate both simple and multiple linear regression models as well as polynomial regression models. I also practiced applying evaluation metrics, including MSE, RMSE and R² Score to measure model performance. RMSE gives us the average prediction error in the same units as the target, while R² score shows how much of the variance in the target variable is explained by the model. These metrics help us understand whether a particular model is accurate and useful.

In [19]:
from google.colab import drive
drive.mount('/content/drive')

!git clone https://github.com/numair-2003/AIML-Internship-NumairFahad.git

Mounted at /content/drive
Cloning into 'AIML-Internship-NumairFahad'...
remote: Enumerating objects: 54, done.
remote: Counting objects: 100% (54/54), done.
remote: Compressing objects: 100% (48/48), done.
remote: Total 54 (delta 22), reused 20 (delta 4), pack-reused 0 (from 0)
Receiving objects: 100% (54/54), 846.30 KiB | 2.68 MiB/s, done.
Resolving deltas: 100% (22/22), done.


In [ ]:
!cp "/content/drive/MyDrive/Colab Notebooks/YourNotebook.ipynb" "/content/your-repo/"